export type ColorBrewerName =
    'OrRd'
    | 'PuBu'
    | 'BuPu'
    | 'Oranges'
    | 'BuGn'
    | 'YlOrBr'
    | 'YlGn'
    | 'Reds'
    | 'RdPu'
    | 'Greens'
    | 'YlGnBu'
    | 'Purples'
    | 'GnBu'
    | 'Greys'
    | 'YlOrRd'
    | 'PuRd'
    | 'Blues'
    | 'PuBuGn'
    | 'Viridis'
    | 'Spectral'
    | 'RdYlGn'
    | 'RdBu'
    | 'PiYG'
    | 'PRGn'
    | 'RdYlBu'
    | 'BrBG'
    | 'RdGy'
    | 'PuOr'
    | 'Set2'
    | 'Accent'
    | 'Set1'
    | 'Set3'
    | 'Dark2'
    | 'Paired'
    | 'Pastel2'
    | 'Pastel1'

return nil
